### PR TITLE
feat: Add prerelease info for new version

### LINKED ISSUE #23: Add prerelease information to the new version
### Feature Request

As an extension to generating a standard semver compatible version number, it would be handy to be able to append prerelease information. e.g.

```
nextVersion(prerelease: 'alpha')
```

would generate something like:

```
1.0.0-alpha
```

### PR DESCRIPTION
**Pull request by philippart-s**
- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [X] Link to relevant issues in GitHub or Jira
- [X] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

fix #23 

Adds the capability to add prerelease informations (like alpha, beta, ...) to the next calculated version.

Done features : 
 - [X] No tag :arrow_right: 0.1.0-alpha
 - [x] 0.1.0 (preRelease: 'alpha') / Major :arrow_right: 1.0.0-alpha
 - [x] 0.1.0 (preRelease: 'alpha') / Minor :arrow_right: 0.2.0-alpha
 - [x] 0.1.0 (preRelease: 'alpha') / Patch :arrow_right: 0.1.1-alpha
 - [x] 1.0.0-alpha  :arrow_right: 1.0.0
 - [x] 0.1.0-alpha  :arrow_right: 0.1.0
 - [x] 0.1.1-alpha  :arrow_right: 0.1.1
 - [x] 0.1.0-alpha (preservePreRelease: true) / Major :arrow_right: 1.0.0-alpha
 - [x] 0.1.0-alpha (preservePreRelease: true) / Minor :arrow_right: 0.2.0-alpha
 - [X] 0.1.0-alpha (preservePreRelease: true) / Patch :arrow_right: 0.1.1-alpha
 - [x] 0.1.0-alpha (incrementPreRelease: true) :arrow_right: 0.1.1-alpha.1
 - [x] 0.1.0-alpha (preRelease: 'beta') / Major :arrow_right: 1.0.0-beta
 - [x] 0.1.0-alpha (preRelease: 'beta') / Minor :arrow_right: 0.2.0-beta
 - [x] 0.1.0-alpha (preRelease: 'beta') / Patch :arrow_right:0.1.1-beta

### CODE REVIEW

**Comment by philippart-s** [comment]
In progress but feel free to add remarks or comment during the work in progress 😉

**Comment by philippart-s** [comment]
@garethjevans & @ADI10HERO Can have your opinion for the tests to add in the class test _JenkinsTest_ for this feature ? 

First possibility : 
- shouldAddPreReleaseInformationPipelineWithNoTags
- shouldAddPreReleaseInformationPipelineWithTags
- shouldAddPreReleaseInformationDeclarativePipelineWithTags
- shouldAddPreReleaseInformationDeclarativePipelineWithTagsInEnvironmentBlock

Second possibility 
Or do you prefer that I write only one test that is focus to the ability of adding a prerelease information never mind the context (tag / no tag, declarative pipeline / scripted pipeline, environment block / direct in step).
In that case I think write one test : _shouldAddPreReleaseInformation_

**Comment by garethjevans** [comment]
Hi @philippart-s, I think it would be fine to just add the `shouldAddPreReleaseInformation` test, we already have tests that work in a pipeline and via a freestyle job so I don't think duplicating them would provide that much extra value

File: README.md [hunk 1]
```diff
@@ -34,6 +35,27 @@ pipeline {
...
 ```
+### Use with prerelease information
+For a `1.0.0` existing version the following code :
+
+```
+pipeline {
+    agent any
+
+    environment {
+        NEXT_VERSION = nextVersion(prerelease: 'alpha')
```
Comment (ADI10HERO) [inline README.md:46]:
```suggestion
        NEXT_VERSION = nextVersion(preRelease: 'alpha')
```

File: src/main/java/io/jenkins/plugins/conventionalcommits/NextVersionStep.java [hunk 2]
```diff
@@ -168,13 +224,36 @@ protected String run() throws Exception {
...
 
-        // based on the commit list, determine how to bump the version
-        Version nextVersion = new ConventionalCommits().nextVersion(currentVersion, commitHistory);
+        Version nextVersion;
+        if (!incrementPreRelease) {
+          // based on the commit list, determine how to bump the version
+          nextVersion =
+              new ConventionalCommits().nextVersion(currentVersion, commitHistory);
+        } else {
+          nextVersion = currentVersion.incrementPreReleaseVersion();
```
Comment (ADI10HERO) [inline src/main/java/io/jenkins/plugins/conventionalcommits/NextVersionStep.java:233]: Regarding this, pre-release version would be added by 1 irrespective of the commits right?

**Review by ADI10HERO — COMMENTED** [review]
Rest Everything looks good to me... Still, I would like to pull and test things locally.

File: src/main/java/io/jenkins/plugins/conventionalcommits/NextVersionStep.java (hunk 2 above)
Comment (philippart-s) [reply src/main/java/io/jenkins/plugins/conventionalcommits/NextVersionStep.java:233]: yes, one of my problem is that I wanted to add _preRelease_ option and the derivated options (_incrementPreRelease_ for example) without modify the _ConventionalCommits_ class (by  adding a parameter to the _nextVersion_ method).

The other thing is if we use the commit message to determine if we also increment major / minor / patch version the combinatorial  are quite many.

File: src/test/java/io/jenkins/plugins/conventionalcommits/JenkinsTest.java [hunk 3]
```diff
@@ -162,4 +164,207 @@ public void shouldAddBuildMetadataInformation() throws Exception {
...
+            true));
+
+    WorkflowRun b = rule.assertBuildStatus(Result.SUCCESS, p.scheduleBuild2(0).get());
+
+    System.out.println(JenkinsRule.getLog(b));
+
+    assertThat(JenkinsRule.getLog(b), containsString("Started"));
+    assertThat(JenkinsRule.getLog(b), containsString("nextVersion"));
+    assertThat(JenkinsRule.getLog(b), containsString("Current Tag is: 0.2.0-alpha"));
+    assertThat(JenkinsRule.getLog(b), containsString("0.2.0\n"));
```
Comment (ADI10HERO) [inline src/test/java/io/jenkins/plugins/conventionalcommits/JenkinsTest.java:367]:
```suggestion
    assertThat(JenkinsRule.getLog(b), containsString("0.2.0"));
```

A test was failing, when I removed the `\n` it passed :)

**Review by ADI10HERO — APPROVED** [review]

File: src/test/java/io/jenkins/plugins/conventionalcommits/JenkinsTest.java (hunk 3 above)
Comment (philippart-s) [reply src/test/java/io/jenkins/plugins/conventionalcommits/JenkinsTest.java:367]: the test must have the _\n_ to test that it's _0.2.0_ and not _0.2.0-alpha_
And the PR checks (with the test) are ok with this test.

**Review by garethjevans — APPROVED** [review]

**Pull request merged**